x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/ShapeQueryPointProcessor.java

```diff
@@ -8,50 +8,272 @@
-        MappedFieldType fieldType = context.getFieldType(fieldName);
-        if (fieldType instanceof PointFieldMapper.PointFieldType == false) {
-            throw new QueryShardException(
-                context,
-                "Expected " + PointFieldMapper.CONTENT_TYPE + " field type for Field [" + fieldName + "] but found " + fieldType.typeName()
+    private static Query buildDisjointQuery(String fieldName, boolean isIndexed, boolean hasDocValues, XYGeometry... luceneGeometries) {
+        // first collect all the documents that contain a shape
+        final BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        if (hasDocValues) {
+            builder.add(new FieldExistsQuery(fieldName), BooleanClause.Occur.FILTER);
+        } else {
+            builder.add(
+                buildIntersectsQuery(
+                    fieldName,
+                    isIndexed,
+                    hasDocValues,
+                    new XYRectangle(-Float.MAX_VALUE, Float.MAX_VALUE, -Float.MAX_VALUE, Float.MAX_VALUE)
+                ),
+                BooleanClause.Occur.FILTER
             );
         }
-        return fieldType.hasDocValues();
+        // then remove all intersecting documents
+        builder.add(buildIntersectsQuery(fieldName, isIndexed, hasDocValues, luceneGeometries), BooleanClause.Occur.MUST_NOT);
+        return builder.build();
+    }
+
+    private static Query buildContainsQuery(String fieldName, boolean isIndexed, boolean hasDocValues, XYGeometry... luceneGeometries) {
+        // for non-point data the result is always false
+        if (allPoints(luceneGeometries) == false) {
+            return new MatchNoDocsQuery();
+        }
+        // for a unique point, it behaves like intersect
+        if (luceneGeometries.length == 1) {
+            return buildIntersectsQuery(fieldName, isIndexed, hasDocValues, luceneGeometries);
+        }
+        // for a multi point, all points needs to be in the document
+        final BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        for (XYGeometry geometry : luceneGeometries) {
+            builder.add(buildIntersectsQuery(fieldName, isIndexed, hasDocValues, geometry), BooleanClause.Occur.FILTER);
+        }
+        return builder.build();
+    }
+
+    private static Query buildWithinQuery(String fieldName, boolean isIndexed, boolean hasDocValues, XYGeometry... luceneGeometries) {
+        final BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        // collect all the intersecting documents
+        builder.add(buildIntersectsQuery(fieldName, isIndexed, hasDocValues, luceneGeometries), BooleanClause.Occur.FILTER);
+        // This is the tricky part as we need to remove all documents that they have at least one disjoint point.
+        // In order to do that, we introduce a InverseXYGeometry which return all documents that have at least one disjoint point
```

Comment:
Very interesting idea! I need to think about it some more, but at first glance it does make sense.